src/dials/util/table_as_hdf5_file.py

```diff
@@ -0,0 +1,271 @@
+    def encode_columns(
+        group: h5py.Group,
+        table: flex.reflection_table,
+        ignore: Optional[List[str]] = None,
+    ) -> None:
+        """Encode the columns of a reflection table."""
+
+        for key, data in table.cols():
+            if ignore and key in ignore:
+                continue
+            if isinstance(data, flex.shoebox):
+                ReflectionListEncoder.encode_shoebox(group, data, key)
+            elif isinstance(data, flex.int6):
+                s = data.size()
+                this_data = flumpy.to_numpy(data.as_int()).reshape(s, 6)
+                group.create_dataset(
+                    key, data=this_data, shape=this_data.shape, dtype=this_data.dtype
+                )
+            elif isinstance(data, flex.std_string):
+                this_data = [s.encode("utf-8") for s in data]
+                group.create_dataset(key, data=this_data, shape=(len(this_data),))
+            else:
+                this_data = flumpy.to_numpy(data)
+                group.create_dataset(
+                    key, data=this_data, shape=this_data.shape, dtype=this_data.dtype
+                )
+
+    @staticmethod
+    def encode_shoebox(group: h5py.Group, data: flex.shoebox, key: str):
+        """Encode a column of shoeboxes."""
+        sbdata, bg, mask = data.get_shoebox_data_arrays()
+        panel = data.panels()
+        bbox = data.bounding_boxes()
+        data = flumpy.to_numpy(sbdata)
+        lz4 = hdf5plugin.LZ4()
+        sbox_group = group.create_group(key)
+        sbox_group.create_dataset(
+            "shoebox_data",
+            data=data,
+            shape=data.shape,
+            dtype=data.dtype,
+            compression=lz4,
+        )
+        data = flumpy.to_numpy(bg)
+        sbox_group.create_dataset(
+            "shoebox_background",
+            data=data,
+            shape=data.shape,
+            dtype=data.dtype,
+            compression=lz4,
+        )
+        data = flumpy.to_numpy(mask)
+        sbox_group.create_dataset(
+            "shoebox_mask",
+            data=data,
+            shape=data.shape,
+            dtype=data.dtype,
+            compression=lz4,
+        )
+        s = bbox.size()
+        this_data = flumpy.to_numpy(bbox.as_int()).reshape(s, 6)
+        sbox_group.create_dataset(
+            "bbox",
+            data=this_data,
+            shape=this_data.shape,
+            dtype=this_data.dtype,
+            compression=lz4,
+        )
+        data = flumpy.to_numpy(panel)
+        sbox_group.create_dataset(
+            "panel", data=data, shape=data.shape, dtype=data.dtype, compression=lz4
+        )
+
+
+class ReflectionListDecoder(object):
+    """Decoder for the reflection data."""
+
+    @staticmethod
+    def decode(handle: h5py.File) -> List[flex.reflection_table]:
+        """Decode the data to a list of reflection tables."""
+
+        # Get the group containing the reflection data
+        g = handle["entry/data_processing"]
```

Comment:
I don't think this should go under `entry/data_processing` since that is a distraction - this is _our_ reflection format not something compliant with e.g. NeXus. We _do_ care about what steps this came from as that impacts the provenance of the data so:

```
/dials/find
/dlals/index
```

_may_ be a more useful group to put things under. We can then include links so that all the "columns" you would expect to find under e.g. `index` are present even if they refer back to results from spot finding.

Reply:
Yes to /dials/ , to include `find` or `index` would not really be possible in the current form of the function call `flex.reflection_table.as_file(filename)` without some complex introspection logic. So maybe `dials/processing` at the moment. Or would you like an extra optional argument for `as_file` which would be a string to use for the name of the second level in the tree, but is only used when saving in hdf5 format?

Reply:
I have changed the top level name to `/dials`. I have left the second level as a generic `/processing` for now, but removed any requirements on the name at the second level, so we could add naming based on process in future.